clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -258,61 +337,75 @@ class device_vector {
   void shrink_to_fit(void) {
     if (_size != capacity()) {
       size_type tmp_capacity = ::std::max(_size, _min_capacity());
-      auto tmp = _alloc.allocate(tmp_capacity);
+      auto tmp =
+          ::std::allocator_traits<Allocator>::allocate(_alloc, tmp_capacity);
       if (_size > 0) {
         ::std::copy(
             oneapi::dpl::execution::make_device_policy(get_default_queue()),
             begin(), end(), tmp);
       }
-      _alloc.deallocate(_storage, _capacity);
+      ::std::allocator_traits<Allocator>::deallocate(_alloc, _storage,
+                                                     _capacity);
       _storage = tmp;
       _capacity = tmp_capacity;
     }
   }
   void assign(size_type n, const T &x) {
     resize(n);
     if (_size > 0) {
-      ::std::fill(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                begin(), begin() + n, x);
+      ::std::fill(
+          oneapi::dpl::execution::make_device_policy(get_default_queue()),
+          begin(), begin() + n, x);
     }
   }
   template <typename InputIterator>
   void
   assign(InputIterator first,
          typename ::std::enable_if<internal::is_iterator<InputIterator>::value,
-                                 InputIterator>::type last) {
+                                   InputIterator>::type last) {
     auto n = ::std::distance(first, last);
     resize(n);
     if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                first, last, begin());
+      ::std::copy(
+          oneapi::dpl::execution::make_device_policy(get_default_queue()),
+          first, last, begin());
     }
   }
-  void clear(void) { _size = 0; }
+  void clear(void) {
+    _destroy(_size);
+    _size = 0;
+  }
   bool empty(void) const { return (size() == 0); }
   void push_back(const T &x) { insert(end(), size_type(1), x); }
   void pop_back(void) {
-    if (_size > 0)
+    if (_size > 0) {
+      _destroy(1, _size - 1);
       --_size;
+    }
   }
   iterator erase(iterator first, iterator last) {
     auto n = ::std::distance(first, last);
     if (last == end()) {
+      _destroy(n, _size - n);
       _size = _size - n;
       return end();
     }
     auto m = ::std::distance(last, end());
     if (m <= 0) {
       return end();
     }
-    auto tmp = _alloc.allocate(m);
+    auto tmp = ::std::allocator_traits<Allocator>::allocate(_alloc, m);
     // copy remainder to temporary buffer.
     ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
```

Comment:
Why copy to a temporary buffer? Just move ahead is much simpler.

Reply:
This is to take advantage of the parallel algorithms, as input and output sequences may overlap if we just move ahead.

I suppose we can check if they overlap and choose which method we use accordingly.  I've done this.